### PR TITLE
feat: Add Flagsmith provider

### PR DESCRIPTION
**Pull request by vpetrusevici**
## This PR
Adds a library project for Flagsmith provider with implementation and tests

### Related Issues
https://github.com/open-feature/dotnet-sdk-contrib/issues/90
https://github.com/Flagsmith/flagsmith-dotnet-client/issues/78

### Notes
We are going to use OpenFeature and Flagsmith in our projects. Unfortunately  dotnet-sdk-contrib doesn't have Flagsmith provider, so i implemented it.

### Follow-up Tasks

- [x] Implement
- [x] Add unit tests for general funcations
- [x] Add unit tests for using OpenFeature evaluation context
- [x] Review from Flagsmith team

### How to test
Just create FlagsmithProvider using FlagsmithClient or FlagsmithConfiguration according to https://github.com/Flagsmith/flagsmith-dotnet-client/tree/main/Example

### CODE REVIEW

**Comment by beeme1mr** [comment]
Hey @dabeeeenster, could someone from the Flagsmith team please take a look when you have a moment?

File: test/OpenFeature.Contrib.Hooks.Otel.Test/OpenFeature.Contrib.Hooks.Otel.Test.csproj [hunk 1]
```diff
@@ -4,9 +4,9 @@
         <ProjectReference Include="..\..\src\OpenFeature.Contrib.Hooks.Otel\OpenFeature.Contrib.Hooks.Otel.csproj" />
     </ItemGroup>
 
-    <ItemGroup>
-      <PackageReference Include="OpenTelemetry" Version="1.4.0" />
-      <PackageReference Include="OpenTelemetry.Exporter.InMemory" Version="1.4.0" />
+    <ItemGroup>
+      <PackageReference Include="OpenTelemetry" Version="1.4.0" />
+      <PackageReference Include="OpenTelemetry.Exporter.InMemory" Version="1.4.0" />
```
Comment (beeme1mr) [inline test/OpenFeature.Contrib.Hooks.Otel.Test/OpenFeature.Contrib.Hooks.Otel.Test.csproj:9]: Please revert this change because it will bump the version of the OTel hook.

Comment (vpetrusevici) [reply test/OpenFeature.Contrib.Hooks.Otel.Test/OpenFeature.Contrib.Hooks.Otel.Test.csproj:9]: looks like VS changed smth in formatin. I'll revert

Comment (toddbaert) [reply test/OpenFeature.Contrib.Hooks.Otel.Test/OpenFeature.Contrib.Hooks.Otel.Test.csproj:9]: @beeme1mr @vpetrusevici I dont think this will impact any releases. This is under the `test` folder. Release please only watches `src`: https://github.com/open-feature/dotnet-sdk-contrib/blob/d8cac7fa6f757f6d62c9648eb249d80528f2b337/release-please-config.json#L5

Comment (vpetrusevici) [reply test/OpenFeature.Contrib.Hooks.Otel.Test/OpenFeature.Contrib.Hooks.Otel.Test.csproj:9]: @toddbaert updated. I hope it's right

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            {
+                return new ResolutionDetails<bool>(flagKey, defaultValue, reason: Reason.Disabled);
+            }
+
+            var stringValue = await flags.GetFeatureValue(flagKey);
+            if (bool.TryParse(stringValue, out var parsedValue))
+            {
+                return new ResolutionDetails<bool>(flagKey, parsedValue);
+            }
+            return new ResolutionDetails<bool>(flagKey, defaultValue, ErrorType.ParseError, Reason.Error, errorMessage: "Failed to parse value in int type");
```
Comment (toddbaert) [inline src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:78]: I recommend throwing [TypeMismatch ](https://github.com/open-feature/dotnet-sdk/blob/main/src/OpenFeature/Error/TypeMismatchException.cs) instead, in which case the SDK will handle the error.

Comment (toddbaert) [inline src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:78]: You might be able to use generics to reduce some duplication here - one function to do the error changes and type conversion, but that's completely up to you.

Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:78]: ok. I'll change it to throwing an exception

Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:78]: I tried to do it, but didn't found a good solution for reusing TryParse, because these types doesn't have an interface for this...

Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:78]: Can't find this exception in 0.5 version. should I update minimal version?

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        {
+            _flagsmithClient = flagsmithClient;
+        }
+
+        private Task<IFlags> GetFlags(EvaluationContext ctx = null)
+        {
+            var key = ctx?.GetValue("targetingKey")?.AsString;
+            return string.IsNullOrEmpty(key)
+                ? _flagsmithClient.GetEnvironmentFlags()
+                : _flagsmithClient.GetIdentityFlags(key, ctx.AsDictionary().Select(x => new Trait(x.Key, x.Value.AsObject) as ITrait).ToList());
```
Comment (toddbaert) [inline src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:55]: I could have missed something, but I don't see tests for this logic.

I would expect to see at least one test that passes in an OpenFeature evaluation context, and asserts it's correctly converted to traits and `GetIdentityFlags` is called.

**Review by toddbaert — CHANGES_REQUESTED** [review]
Thank you for this addition! I made a few comments, but this looks like a great addition!

I'm marking as "request changes" for now, primarily because I'd like an approval from a representative from Flagsmith first.

Once we have identified a Flagsmith representative, I will add you and them to https://github.com/open-feature/dotnet-sdk-contrib/blob/main/.github/component_owners.yml as part of this PR.

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 3 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:55]: You are right, i didn't add tests for this. I tried to replicate logic from java sdk, but have no idea how to create context for tests :(

File: DotnetSdkContrib.sln [hunk 4]
```diff
@@ -5,28 +5,29 @@ VisualStudioVersion = 17.0.31903.59
...
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenFeature.Contrib.Providers.GOFeatureFlag.Test", "test\OpenFeature.Contrib.Providers.GOFeatureFlag.Test\OpenFeature.Contrib.Providers.GOFeatureFlag.Test.csproj", "{4041B63F-9CF6-4886-8FC7-BD1A7E45F859}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenFeature.Contrib.Providers.GOFeatureFlag.Test", "test\OpenFeature.Contrib.Providers.GOFeatureFlag.Test\OpenFeature.Contrib.Providers.GOFeatureFlag.Test.csproj", "{4041B63F-9CF6-4886-8FC7-BD1A7E45F859}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenFeature.Contrib.Providers.Flagsmith", "src\OpenFeature.Contrib.Providers.Flagsmith\OpenFeature.Contrib.Providers.Flagsmith.csproj", "{47008BEE-7888-4B9B-8884-712A922C3F9B}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenFeature.Contrib.Providers.Flagsmith.Test", "test\OpenFeature.Contrib.Providers.Flagsmith.Test\OpenFeature.Contrib.Providers.Flagsmith.Test.csproj", "{C3BA23C2-BEC3-4683-A64A-C914C3D8037E}"
```
Comment (toddbaert) [inline DotnetSdkContrib.sln:24]: These GUIDs seem to be changing the project type as mentioned here: https://github.com/dotnet/project-system/blob/main/docs/opening-with-new-project-system.md#project-type-guids.

I believe this is Visual Studio specific, so I'm not 100% confident in the change (I don't use Visual Studio-proper)

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 2 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: Changed. But also changed version to 1.3.1 (only for this new package)

Comment (toddbaert) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: > Can't find this exception in 0.5 version. should I update the minimal version?

Yes. I would make the min version 1.0

File: DotnetSdkContrib.sln (hunk 4 above)
Comment (vpetrusevici) [reply DotnetSdkContrib.sln:24]: it should affect if routes wasn't changed

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 2 above)
Comment (toddbaert) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: You could try to extract `TryParse` into a lambda which you pass to the single generic function, but again this is up to you. The duplication isn't that bad IMHO.

File: test/OpenFeature.Contrib.Hooks.Otel.Test/OpenFeature.Contrib.Hooks.Otel.Test.csproj (hunk 1 above)
Comment (toddbaert) [reply test/OpenFeature.Contrib.Hooks.Otel.Test/OpenFeature.Contrib.Hooks.Otel.Test.csproj:9]: I still see whitespace changes but I'm not worried about them. They should not impact anything, since they are not in the `src/` dir.

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 3 above)
Comment (toddbaert) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:57]: If you mean building an OpenFeature evaluation context, it's just a simple "bag" for keys/values. You could just add a few properties to it, and see if they are mapped correctly. [Here's an example test](https://github.com/open-feature/dotnet-sdk/blob/2cbdba80d836f8b7850e8dc5f1f1790ef2ed1aca/test/OpenFeature.Tests/OpenFeatureEvaluationContextTests.cs#L53-L59) which creates a context in the dotnet SDK.

I hope that helps. Let me know if it doesn;'t.

File: DotnetSdkContrib.sln (hunk 4 above)
Comment (toddbaert) [reply DotnetSdkContrib.sln:24]: > it should affect if routes wasn't changed

I'm sorry, I don't understand :sweat_smile:

Comment (vpetrusevici) [reply DotnetSdkContrib.sln:24]: Sorry :)
These changes should not affect anything if the paths in the file have not changed.

**Comment by vpetrusevici** [comment]
I did thm strange. Sorry. I'll recreate the branch

**Comment by vpetrusevici** [comment]
Recreated here https://github.com/open-feature/dotnet-sdk-contrib/pull/91

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 2 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: done

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 3 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:57]: I can try reflection

Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:57]: added!

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 2 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: done :)

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+
+        /// <inheritdoc/>
+        public override Metadata GetMetadata() => Metadata;
+
+        /// <inheritdoc/>
+
+        public override Task<ResolutionDetails<bool>> ResolveBooleanValue(string flagKey, bool defaultValue, EvaluationContext context = null)
+            => ResolveValue(flagKey, defaultValue, bool.TryParse, context);
```
Comment (matthewelwell) [inline src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:93]: Because of the fact that Flagsmith flags can have an enabled state, and a value, in the other Flagsmith providers we use an additional bit of config to allow users of the provider to state if they want to use the value or the state when resolving boolean values. See an example in the Java provider [here](https://github.com/open-feature/java-sdk-contrib/blob/main/providers/flagsmith/src/main/java/dev.openfeature.contrib.providers.flagsmith/FlagsmithProvider.java#L53-L56). 

Would it be possible to do something similar here please?

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                result = null;
+            }
+            return result is not null;
+        }
+
+        /// <summary>
+        ///     convertValue is converting the object return by the proxy response in the right type.
+        /// </summary>
+        /// <param name="node">The value we have received</param>
+        /// <returns>A converted object</returns>
```
Comment (matthewelwell) [inline src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:131]: I'm not sure what we mean by `the proxy response` here? Would the following make more sense? 

```suggestion
        /// <summary>
        ///     convertValue is converting the dynamically typed object received from Flagsmith into the correct type
        /// </summary>
        /// <param name="node">The dynamically typed value we received from Flagsmith</param>
        /// <returns>A correctly typed object representing the flag value</returns>
```

File: test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+            ApiUrl = "https://edge.api.flagsmith.com/api/v1/",
+            EnvironmentKey = string.Empty,
+            EnableClientSideEvaluation = false,
+            EnvironmentRefreshIntervalSeconds = 60,
+            EnableAnalytics = false,
+            Retries = 1
+        };
+
+        [Fact]
+        public void CreateFlagmithProvider_WithValidCredetials_CreatesInstanceSuccessfully()
```
Comment (matthewelwell) [inline test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs:28]:
```suggestion
        public void CreateFlagmithProvider_WithValidCredentials_CreatesProviderInstanceSuccessfully()
```

File: test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+            // Act
+            var flagsmithProvider = new FlagsmithProvider(config);
+
+
+            // Assert
+            Assert.NotNull(flagsmithProvider._flagsmithClient);
+        }
+
+        [Fact]
+        public void CreateFlagmithProvider_WithValidCredetialsAndCustomHttpClient_CreatesInstanceSuccessfully()
```
Comment (matthewelwell) [inline test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs:42]:
```suggestion
        public void CreateFlagmithProvider_WithValidCredentialsAndCustomHttpClient_CreatesProviderInstanceSuccessfully()
```

File: test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+
+            // Assert
+            Assert.Equal(-32.22, result.Value);
+            Assert.Equal("example-feature", result.FlagKey);
+            Assert.Equal(Reason.Disabled, result.Reason);
+            Assert.Equal(ErrorType.None, result.ErrorType);
+        }
+
+        [Fact]
+        public async Task GetDoubleValue_ForEnabledFeatureWithWrongFormatValue_ReturnDefaultValue()
```
Comment (matthewelwell) [inline test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs:193]:
```suggestion
        public async Task GetDoubleValue_ForEnabledFeatureWithWrongFormatValue_ThrowsTypeMismatch()
```

File: test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+
+            // Assert
+            Assert.Equal(-32, result.Value);
+            Assert.Equal("example-feature", result.FlagKey);
+            Assert.Equal(Reason.Disabled, result.Reason);
+            Assert.Equal(ErrorType.None, result.ErrorType);
+        }
+
+        [Fact]
+        public async Task GetIntValue_ForEnabledFeatureWithWrongFormatValue_ReturnDefaultValue()
```
Comment (matthewelwell) [inline test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs:301]:
```suggestion
        public async Task GetIntValue_ForEnabledFeatureWithWrongFormatValue_ThrowsTypeMismatch()
```

File: test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+
+            // Assert
+            Assert.Equal(defaultObject, result.Value);
+            Assert.Equal("example-feature", result.FlagKey);
+            Assert.Equal(Reason.Disabled, result.Reason);
+            Assert.Equal(ErrorType.None, result.ErrorType);
+        }
+
+        [Fact]
+        public async Task GetStructureValue_ForEnabledFeatureWithWrongFormatValue_ReturnDefaultValue()
```
Comment (matthewelwell) [inline test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs:409]:
```suggestion
        public async Task GetStructureValue_ForEnabledFeatureWithWrongFormatValue_ThrowsTypeMismatch()
```

File: test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+
+            var defaultObject = new Value("default");
+            var flagsmithProvider = new FlagsmithProvider(flagsmithClient);
+
+            // Act and Assert
+            await Assert.ThrowsAsync<TypeMismatchException>(() => flagsmithProvider.ResolveStructureValue("example-feature", defaultObject));
+        }
+    }
+
+    public class ExampleConfig
```
Comment (matthewelwell) [inline test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs:426]: I can't see where this class is actually being used?

**Review by matthewelwell — COMMENTED** [review]
Thanks for the effort on this one @vpetrusevici! I've added a few fairly minor comments. There's only one that requires some logic changes.

**Comment by matthewelwell** [comment]
@beeme1mr, I've added my review here and I'm happy being added to the component owners, but could you add another one of the Flagsmith team, [novakzaballa](https://github.com/novakzaballa), as a component owner too please?

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 5 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:93]: sure. i'll try)

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 6 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:131]: I just copied some desctiptions and forgot to change :)

File: test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs (hunk 12 above)
Comment (vpetrusevici) [reply test/OpenFeature.Contrib.Providers.Flagsmith.Test/FlagsmithProviderTest.cs:426]: I just forgot to delete. Thanks)

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 5 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: done!

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        /// </summary>
+        public string TargetingKey { get; set; } = "targetingKey";
+
+        /// <summary>
+        /// Determines whether to resolve a feature value as a boolean or use
+        /// the isFeatureEnabled as the flag itself. These values will be false
+        /// and true respectively.
+        /// Default: false
+        /// </summary>
+        public bool UsingBooleanConfigValue { get; set; }
```
Comment (matthewelwell) [inline src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs:36]: I think I'd prefer to see this in it's own config class? Perhaps we can create a `FlagsmithProviderConfiguration` class which extends the `FlagsmithConfiguration` class from the Flagsmith lib?

**Review by matthewelwell — COMMENTED** [review]
Thanks @vpetrusevici, just one extra comment.

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 13 above)
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: done, but i created a separate one. Just to avoid additional dependency

**Comment by vpetrusevici** [comment]
Hi @beeme1mr,
Do we need anything else?

**Review by matthewelwell — APPROVED** [review]
Looks good from a Flagsmith perspective.

File: src/OpenFeature.Contrib.Providers.Flagsmith/openfeature-icon.png
Comment (beeme1mr) [inline src/OpenFeature.Contrib.Providers.Flagsmith/openfeature-icon.png:1]: Is this icon being used?

**Review by beeme1mr — COMMENTED** [review]
I just had one question about the icon, but the rest looks good. @toddbaert, did you have any other concerns?

Before we merge, could you lease update the `component_owners.yml` to include @matthewelwell, [novakzaballa](https://github.com/novakzaballa), and yourself?

File: src/OpenFeature.Contrib.Providers.Flagsmith/openfeature-icon.png
Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/openfeature-icon.png:1]: yep. i just copied from another provider

Comment (vpetrusevici) [reply src/OpenFeature.Contrib.Providers.Flagsmith/openfeature-icon.png:1]: sorry. no. i should delete it)

**Comment by vpetrusevici** [comment]
> I just had one question about the icon, but the rest looks good. @toddbaert, did you have any other concerns?
> 
> Before we merge, could you lease update the `component_owners.yml` to include @matthewelwell, [novakzaballa](https://github.com/novakzaballa), and yourself?

done

**Comment by beeme1mr** [comment]
Could you please also add a readme under `src/OpenFeature.Contrib.Providers.Flagsmith/`? It should have installation instructions and usage examples. Sorry, I didn't notice that it was missing in my last review!

**Comment by vpetrusevici** [comment]
> Could you please also add a readme under `src/OpenFeature.Contrib.Providers.Flagsmith/`? It should have installation instructions and usage examples. Sorry, I didn't notice that it was missing in my last review!

I just copied readme form flagd provider and adapted to Flagsmith provider.

File: src/OpenFeature.Contrib.Providers.Flagsmith/README.md [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+```
+### Package Reference
+
+```xml
+<PackageReference Include="OpenFeature.Contrib.Providers.Flagsmith" />
+```
+### Packet cli
+
+```shell
+paket add OpenFeature.Contrib.Providers.Flagsmith
```
Comment (beeme1mr) [inline src/OpenFeature.Contrib.Providers.Flagsmith/README.md:28]:
```suggestion
packet add OpenFeature.Contrib.Providers.Flagsmith
```
This looks like a typo but I'm not familiar with packet.

File: src/OpenFeature.Contrib.Providers.Flagsmith/README.md [hunk 15]
```diff
@@ -0,0 +1,114 @@
+# Flagsmith Feature Flag .NET Provider
+
+The Flagsmith Flag provider allows you to connect to your Flagsmith instance.
```
Comment (beeme1mr) [inline src/OpenFeature.Contrib.Providers.Flagsmith/README.md:3]:
```suggestion
The Flagsmith provider allows you to connect to your Flagsmith instance through the OpenFeature SDK
```

File: src/OpenFeature.Contrib.Providers.Flagsmith/README.md [hunk 16]
```diff
@@ -0,0 +1,114 @@
+# Flagsmith Feature Flag .NET Provider
```
Comment (beeme1mr) [inline src/OpenFeature.Contrib.Providers.Flagsmith/README.md:1]:
```suggestion
# Flagsmith .NET Provider
```

**Review by beeme1mr — APPROVED** [review]

**Comment by vpetrusevici** [comment]
@toddbaert can we complete the PR? :)

File: .github/component_owners.yml [hunk 17]
```diff
@@ -9,6 +9,10 @@ components:
     - toddbaert
   src/OpenFeature.Contrib.Providers.GOFeatureFlag:
     - thomaspoignant
+  src/OpenFeature.Contrib.Providers.Flagsmith:
+    - vpetrusevici
+    - matthewelwell
+    - novakzaballa
```
Comment (toddbaert) [inline .github/component_owners.yml:15]: Not a blocker, but `novakzaballa` will not be added to reviews since they are not a member of the org. They can create a PR to add themselves [here](https://github.com/open-feature/community/blob/ab73778324513cc789f184408afcf2912fceec53/config/open-feature/org.yaml#L29).

File: src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs (hunk 2 above)
Comment (toddbaert) [reply src/OpenFeature.Contrib.Providers.Flagsmith/FlagsmithProvider.cs]: Nice!

**Comment by vpetrusevici** [comment]
> @vpetrusevici very nice! I like the delegate implementation. This is a great addition!
> 
> I'm approving but you have some minor formatting issues (you should be able to test your fixes with `dotnet-format --folder --check`). I'll keep an eye on the PR and re-run the checks / merge as soon as they pass.
I added ```#pragma warning disable format``` for json in tests. it's strange but dotnet-format tries to fix this string...

File: .github/component_owners.yml (hunk 17 above)
Comment (vpetrusevici) [reply .github/component_owners.yml]: removed
